cardano-crypto-class/src/Cardano/Crypto/DSIGN/SchnorrSecp256k1.hs

```diff
@@ -0,0 +1,296 @@
+import System.IO.Unsafe (unsafeDupablePerformIO, unsafePerformIO)
+import Cardano.Binary (FromCBOR (fromCBOR), ToCBOR (toCBOR, encodedSizeExpr))
+import Foreign.Ptr (Ptr, castPtr, nullPtr)
+import Foreign.C.Types (CUChar)
+import Foreign.Marshal.Alloc (allocaBytes)
+import Cardano.Crypto.Schnorr (
+  SECP256k1XOnlyPubKey, 
+  secpKeyPairCreate, 
+  SECP256k1Context, 
+  secpKeyPairXOnlyPub,
+  SECP256k1SecKey, 
+  secpSchnorrSigVerify, 
+  secpContextSignVerify,
+  SECP256k1SchnorrSig, 
+  secpSchnorrSigSignCustom, 
+  secpContextCreate
+  )
+import NoThunks.Class (NoThunks, OnlyCheckWhnfNamed (OnlyCheckWhnfNamed))
+import Cardano.Crypto.DSIGN.Class (
+  DSIGNAlgorithm (VerKeyDSIGN, 
+                  SignKeyDSIGN, 
+                  SigDSIGN,
+                  SeedSizeDSIGN, 
+                  SizeSigDSIGN, 
+                  SizeSignKeyDSIGN, 
+                  SizeVerKeyDSIGN, 
+                  algorithmNameDSIGN,
+                  deriveVerKeyDSIGN, 
+                  signDSIGN, 
+                  verifyDSIGN, 
+                  genKeyDSIGN, 
+                  rawSerialiseSigDSIGN,
+                  Signable, 
+                  rawSerialiseVerKeyDSIGN, 
+                  rawSerialiseSignKeyDSIGN, 
+                  rawDeserialiseVerKeyDSIGN,
+                  rawDeserialiseSignKeyDSIGN, 
+                  rawDeserialiseSigDSIGN), 
+  encodeVerKeyDSIGN, 
+  encodedVerKeyDSIGNSizeExpr, 
+  decodeVerKeyDSIGN, 
+  encodeSignKeyDSIGN, 
+  encodedSignKeyDESIGNSizeExpr, 
+  decodeSignKeyDSIGN, 
+  encodeSigDSIGN, 
+  encodedSigDSIGNSizeExpr, 
+  decodeSigDSIGN
+  )
+import Cardano.Crypto.Util (SignableRepresentation (getSignableRepresentation))
+
+data SchnorrSecp256k1DSIGN
+
+instance DSIGNAlgorithm SchnorrSecp256k1DSIGN where
+  type SeedSizeDSIGN SchnorrSecp256k1DSIGN = 32
+  type SizeSigDSIGN SchnorrSecp256k1DSIGN = 64
+  type SizeSignKeyDSIGN SchnorrSecp256k1DSIGN = 32
+  type SizeVerKeyDSIGN SchnorrSecp256k1DSIGN = 64
+  type Signable SchnorrSecp256k1DSIGN = SignableRepresentation
+  newtype VerKeyDSIGN SchnorrSecp256k1DSIGN = 
+    VerKeySchnorr256k1 (ForeignPtr SECP256k1XOnlyPubKey)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "VerKeySchnorr256k1" (ForeignPtr SECP256k1XOnlyPubKey)
+        )
+  newtype SignKeyDSIGN SchnorrSecp256k1DSIGN = 
+    SignKeySchnorr256k1 (ForeignPtr SECP256k1SecKey)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "SignKeySchnorr256k1" (ForeignPtr SECP256k1SecKey)
+        )
+  newtype SigDSIGN SchnorrSecp256k1DSIGN = 
+    SigSchnorr256k1 (ForeignPtr SECP256k1SchnorrSig)
+      deriving NoThunks via (
+        OnlyCheckWhnfNamed "SigSchnorr256k1" (ForeignPtr SECP256k1SchnorrSig)
+        )
+  algorithmNameDSIGN _ = "schnorr-secp256k1"
+  {-# NOINLINE deriveVerKeyDSIGN #-}
+  deriveVerKeyDSIGN (SignKeySchnorr256k1 fp) = 
+    unsafeDupablePerformIO . withForeignPtr fp $ \skp -> do
+      let skp' :: Ptr CUChar = castPtr skp
+      allocaBytes 96 $ \kpp -> do
```

Comment:
I note that this package has some utilities for slightly safer allocation, in [here](https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Foreign.hs). I think you can use this in your foreign declarations (like [here](https://github.com/input-output-hk/cardano-base/blob/39f9cf0eaf15a6888a47ac9d8de0b37dd1131d0b/cardano-crypto-class/src/Cardano/Crypto/Libsodium/C.hs#L90)) to get some more safety and fewer magic numbers.

Reply:
I saw that, but I actually don't think it helps all that much. I prefer keeping close to C, magic numbers and all, and avoiding the cost of 'calling down' type-level numbers to the value level.

Reply:
Nonetheless, if it's going to be in this repository, I think it should follow the practices of this repository :shrug:

Reply:
I respectfully disagree. You're injecting quite expensive code into (arguably) a hot code path for no real benefit, since nobody outside of the maintainers of this repo will see it. If magic numbers are a concern, I am happy to replace those with named constants.

Reply:
> You're injecting quite expensive code into (arguably) a hot code path for no real benefit

I think both of these are wrong:
- The code is not expensive. GHC happily reifies the statically known information at compile time: I looked at the Core for the `Ed25519` module and it has numeric literals as the size arguments.
- This is not a hot path: we are calling expensive primitives, we shouldn't need to worry too much about the FFI overhead.

In any case, please remove the magic numbers. You already had to define named constants in the class: use them. I also don't think there's a good reason not to use the library helpers here, and it makes things more consistent.